### PR TITLE
Prevent TieredPGO inlining from breaking stack trace walking logic

### LINKED ISSUE #2617: [BUG] Stack inspection done by playwright is fragile and breaks with Dynamic PGO enabled
**Context:**

We enabled DynamicPGO in .NET 8 Preview 5 and are seeing reports of some failures in applications that use playwright.

See for instance https://github.com/dotnet/runtime/issues/87628

This should readily repro if you run release mode builds with .NET 8 Preview 5 and default runtime options.

**Describe the bug**

https://github.com/microsoft/playwright-dotnet/blob/ca7b02ac910e573666f48d502bfabb1c17639e68/src/Playwright/Transport/Connection.cs#L477-L502

is looking for a specific pattern of method names in the stack trace. This is fragile unless all the methods being looked for are marked with `[MethodImpl(MethodImplOptions.NoInlining)]`. Note that with PGO the JIT may now inline methods at virtual, interface, and delegate call sites.

I suspect the fix may be as simple as marking this method noinline

https://github.com/microsoft/playwright-dotnet/blob/ca7b02ac910e573666f48d502bfabb1c17639e68/src/Playwright/Transport/Connection.cs#L521-L528

but there may be more cases elsewhere in this code that need similar treatment.

### PR DESCRIPTION
**Pull request by slang25**
Fixes #2617 

I've recreated the issue locally on macOS on my M2 MBP, this fixes the exception, I need to check that the traces have the right API names captured still.

### CODE REVIEW

File: src/Playwright/Transport/Connection.cs [hunk 1]
```diff
@@ -536,6 +538,7 @@ private static bool IsPlaywrightInternalNamespace(string namespaceName)
             namespaceName.StartsWith("Microsoft.Playwright.Helpers", StringComparison.InvariantCultureIgnoreCase));
     }
 
+    [MethodImpl(MethodImplOptions.NoInlining)] // This method is also a stacktrace marker.
```
Comment (slang25) [inline src/Playwright/Transport/Connection.cs:541]: While not strictly necessary, it seems a good idea to make this explicit

**Comment by slang25** [comment]
Unfortunately I've done some further testing and it looks like a lot of the method names captured in the trace will change to base types, which isn't desired, I'll look to improve this PR.

**Comment by slang25** [comment]
Update:

So this PR fixes the `NullReferenceException`, but in order to get the right API name captured like this:
https://github.com/microsoft/playwright-dotnet/blob/ca7b02ac910e573666f48d502bfabb1c17639e68/src/Playwright.Tests/TracingTests.cs#L69-L75
we're going to need to sprinkle `NoInlining` over all of the public APIs. I've been thinking about other approaches that don't do stack frame walking, however some public APIs flow through other public APIs , for example `Page.GotoAsync → Frame.GotoAsync` ([see here](https://github.com/microsoft/playwright-dotnet/blob/ca7b02ac910e573666f48d502bfabb1c17639e68/src/Playwright/Core/Page.cs#L292-L293)), which makes that harder to do in a non-breaking way.

**Comment by slang25** [comment]
So this is now working and tested with some traces. Unfortunately the cleanest way I could think to do this was add `[MethodImpl(MethodImplOptions.NoInlining)]` absolutely everywhere on the public APIs of anything that inherits `ChannelOwnerBase`, this is enforced with a convention test.

File: src/Playwright.Tests/ConventionTests.cs [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+        foreach (var type in types)
+        {
+            var methods = type
+                .GetMethods(BindingFlags.Public | BindingFlags.Instance | BindingFlags.DeclaredOnly)
+                .Where(m => !m.IsSpecialName)
+                .Where(m => m.Name != "Dispose" && m.Name != "DisposeAsync" && m.Name != "ToString");
+
+            foreach (var method in methods)
+            {
+                Assert.True(method.MethodImplementationFlags.HasFlag(MethodImplAttributes.NoInlining), $"{type.Name}.{method.Name} is not marked with [MethodImpl(MethodImplOptions.NoInlining)]");
```
Comment (martincostello) [inline src/Playwright.Tests/ConventionTests.cs:28]: If you put these messages into a list and then assert the list is empty, you could fail the test with everything that needs to be fixed at once in a single message that concatenates them.

Comment (slang25) [reply src/Playwright.Tests/ConventionTests.cs:28]: Sure, tell me now after I reran it 50 times!

File: src/Playwright.Tests/ConventionTests.cs [hunk 3]
```diff
@@ -0,0 +1,47 @@
+using System.Globalization;
+using System.Reflection;
+using System.Text;
+
+namespace Microsoft.Playwright.Tests;
+
+public class ConventionTests
+{
+    // To ensure that public method are not inlined by the new tiered PGO JIT mode,
+    // we need to mark them with [MethodImpl(MethodImplOptions.NoInlining)]
```
Comment (mxschmitt) [inline src/Playwright.Tests/ConventionTests.cs:10]:
```suggestion
    // we need to mark them with [MethodImpl(MethodImplOptions.NoInlining)]
    // See https://github.com/microsoft/playwright-dotnet/issues/2617
```

**Comment by mxschmitt** [comment]
@slang25 thank you so much for looking into this and @AndyAyersMS for filing it inside this repo with all the details!

**Comment by slang25** [comment]
My pleasure @mxschmitt 🙂

**Pull request merged**